orchagent/pfcwdorch.cpp

```diff
@@ -0,0 +1,755 @@
+void PfcWdSwOrch<DropHandler, ForwardHandler>::endWatchdogThread(void)
+{
+    SWSS_LOG_ENTER();
+
+    if (!m_runPfcWdSwOrchThread.load())
+    {
+        return;
+    }
+
+    m_runPfcWdSwOrchThread = false;
+
+    m_cvSleep.notify_all();
+
+    if (m_pfcWatchdogThread != nullptr)
+    {
+        SWSS_LOG_INFO("Wait for PFC Watchdog thread to end");
+
+        m_pfcWatchdogThread->join();
+    }
+
+    SWSS_LOG_INFO("PFC Watchdog thread ended");
+}
+
+template <typename DropHandler, typename ForwardHandler>
+PfcDurationWatchdog<DropHandler, ForwardHandler>::PfcDurationWatchdog(
+        DBConnector *db, vector<string> &tableNames):
+    PfcWdSwOrch<DropHandler, ForwardHandler>(db, tableNames)
+{
+    SWSS_LOG_ENTER();
+}
+
+template <typename DropHandler, typename ForwardHandler>
+PfcDurationWatchdog<DropHandler, ForwardHandler>::~PfcDurationWatchdog(void)
+{
+    SWSS_LOG_ENTER();
+}
+
+template <typename DropHandler, typename ForwardHandler>
+vector<sai_port_stat_t> PfcDurationWatchdog<DropHandler, ForwardHandler>::getPortCounterIds(
+        sai_object_id_t queueId)
+{
+    SWSS_LOG_ENTER();
+
+    static const vector<sai_port_stat_t> PfcDurationIdMap =
+    {
+        SAI_PORT_STAT_PFC_0_RX_PAUSE_DURATION,
+        SAI_PORT_STAT_PFC_1_RX_PAUSE_DURATION,
+        SAI_PORT_STAT_PFC_2_RX_PAUSE_DURATION,
+        SAI_PORT_STAT_PFC_3_RX_PAUSE_DURATION,
+        SAI_PORT_STAT_PFC_4_RX_PAUSE_DURATION,
+        SAI_PORT_STAT_PFC_5_RX_PAUSE_DURATION,
+        SAI_PORT_STAT_PFC_6_RX_PAUSE_DURATION,
+        SAI_PORT_STAT_PFC_7_RX_PAUSE_DURATION,
+    };
+
+    static const vector<sai_port_stat_t> PfcRxPktsIdMap =
+    {
+        SAI_PORT_STAT_PFC_0_RX_PKTS,
+        SAI_PORT_STAT_PFC_1_RX_PKTS,
+        SAI_PORT_STAT_PFC_2_RX_PKTS,
+        SAI_PORT_STAT_PFC_3_RX_PKTS,
+        SAI_PORT_STAT_PFC_4_RX_PKTS,
+        SAI_PORT_STAT_PFC_5_RX_PKTS,
+        SAI_PORT_STAT_PFC_6_RX_PKTS,
+        SAI_PORT_STAT_PFC_7_RX_PKTS,
+    };
+
+    sai_attribute_t attr;
+    attr.id = SAI_QUEUE_ATTR_INDEX;
+
+    sai_status_t status = sai_queue_api->get_queue_attribute(queueId, 1, &attr);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get queue index 0x%lx: %d", queueId, status);
+        return { };
+    }
+
+    size_t index = attr.value.u8;
+    vector<sai_port_stat_t> portStatIds =
+    {
+        PfcDurationIdMap[index],
+        PfcRxPktsIdMap[index],
+    };
+
+    auto commonIds = PfcWdOrch<DropHandler, ForwardHandler>::getPortCounterIds(queueId);
+    portStatIds.insert(portStatIds.end(), commonIds.begin(), commonIds.end());
+
+    return move(portStatIds);
+}
+
+template <typename DropHandler, typename ForwardHandler>
+vector<sai_queue_stat_t> PfcDurationWatchdog<DropHandler, ForwardHandler>::getQueueCounterIds(
+        sai_object_id_t queueId)
+{
+    SWSS_LOG_ENTER();
+
+    vector<sai_queue_stat_t> queueStatIds =
+    {
+        SAI_QUEUE_STAT_CURR_OCCUPANCY_BYTES,
```

Comment:
this could also be different for different platforms, for example, some platform use the pkts instead of bytes.